tests/django_pg_migration_tools/test_operations.py

```diff
@@ -427,3 +454,355 @@ def test_when_not_allowed_to_migrate(self):
+            cursor.execute(_SET_INDEX_INVALID, {"index_name": "unique_int_field_idx"})
+            # Also, set the lock_timeout to check it has been returned to
+            # its original value once the unique index creation is completed.
+            cursor.execute(_SET_LOCK_TIMEOUT)
+
+        # Prove that the invalid unique index exists before the operation runs:
+        with connection.cursor() as cursor:
+            cursor.execute(
+                operations.SaferAddUniqueConstraint.CHECK_INVALID_INDEX_QUERY,
+                {"index_name": "unique_int_field_idx"},
+            )
+            assert cursor.fetchone()
+
+        # Prove that the constraint does **not** already exist.
+        with connection.cursor() as cursor:
+            cursor.execute(
+                operations.SaferAddUniqueConstraint._CHECK_EXISTING_CONSTRAINT_QUERY,
+                {"constraint_name": "unique_int_field"},
+            )
+            assert not cursor.fetchone()
+
+        project_state = ProjectState()
+        project_state.add_model(ModelState.from_model(IntModel))
+        new_state = project_state.clone()
+
+        operation = operations.SaferAddUniqueConstraint(
+            model_name="intmodel",
+            constraint=UniqueConstraint(
+                fields=("int_field",),
+                name="unique_int_field",
+            ),
+        )
+
+        assert operation.describe() == (
+            "Concurrently adds a UNIQUE index unique_int_field_idx on model intmodel "
+            "on field(s) ('int_field',) if the index does not exist. Then, adds the "
+            "constraint using the just-created index. NOTE: "
+            "Using django_pg_migration_tools SaferAddUniqueConstraint operation."
+        )
+
+        name, args, kwargs = operation.deconstruct()
+        assert name == "SaferAddUniqueConstraint"
+        assert args == []
+        assert kwargs == {"model_name": "intmodel", "constraint": operation.constraint}
+
+        operation.state_forwards(self.app_label, new_state)
+        assert (
+            len(new_state.models[self.app_label, "intmodel"].options["constraints"])
+            == 1
+        )
+        assert (
+            new_state.models[self.app_label, "intmodel"].options["constraints"][0].name
+            == "unique_int_field"
+        )
+
+        # Proceed to add the unique index followed by the constraint:
+        with connection.schema_editor(atomic=False, collect_sql=False) as editor:
+            with utils.CaptureQueriesContext(connection) as queries:
+                operation.database_forwards(
+                    self.app_label, editor, project_state, new_state
+                )
+
+        # Assert the index has disappeared!!!
+        # Be no alarmed, this seems a bit strange, but Postgres has "swapped"
+        # the index for a constraint instead. Keeping both the index and the
+        # constraint would result in unnecessarily occupied space.
```

Comment:
🐼  I don't think unnecessary space usage is the issue. Rather, when the constraint is created using the index, the index becomes owned by the constraint, and therefore cannot be manipulated independently as an index anymore. The way to enforce that is to remove it from the catalog of indexes — i.e., it disappears.

Reply:
I've removed this comment as part of https://github.com/kraken-tech/django-pg-migration-tools/commit/b3c627119240240411b7382bcf65c5c8a8896cdc